File: packages/db/src/repositories/promotion-run.repository.ts
import { PersistRunStateInputSchema } from '@ff-promo/contracts';
import { z } from 'zod';
import type { PrismaClient, PromotionStatus } from '../../generated/client/index.js';

const PromotionRunCreateSchema = z.object({
  pipelineId: z.string(),
  flagKey: z.string(),
});

export class PromotionRunRepository {
  constructor(private readonly db: PrismaClient) {}

  async create(input: { pipelineId: string; flagKey: string }) {
    const { pipelineId, flagKey } = PromotionRunCreateSchema.parse(input);

    const pipeline = await this.db.pipeline.findUniqueOrThrow({
      where: { id: pipelineId },
    });

    return this.db.promotionRun.create({
      data: {
        pipelineId,
        flagKey,
        pipelineVersion: pipeline.version,
      },
    });
  }

  async updateState(input: {
    promotionRunId: string;
    status: PromotionStatus;
    currentStageIndex?: number;
    pauseReason?: string;
    temporalWorkflowId?: string;
  }) {
    const parsed = PersistRunStateInputSchema.parse({
      promotionRunId: input.promotionRunId,
      status: input.status,
      currentStageIndex: input.currentStageIndex,
      pauseReason: input.pauseReason,
    });

    const existing = await this.db.promotionRun.findUniqueOrThrow({
      where: { id: parsed.promotionRunId },
    });

    const isFirstActiveTransition =
      existing.status !== 'active' && parsed.status === 'active';
    const temporalWorkflowId =
      input.temporalWorkflowId ??
      (isFirstActiveTransition && !existing.temporalWorkflowId
        ? parsed.promotionRunId
        : undefined);

    return this.db.promotionRun.update({
      where: { id: parsed.promotionRunId },
      data: {
        status: parsed.status,
        ...(parsed.currentStageIndex !== undefined && {
          currentStageIndex: parsed.currentStageIndex,
        }),
        ...(parsed.pauseReason !== undefined && {
          pauseReason: parsed.pauseReason,
        }),
        ...(temporalWorkflowId !== undefined && { temporalWorkflowId }),
      },
    });
  }

  async findById(id: string) {
    return this.db.promotionRun.findUnique({
      where: { id },
    });
  }

  async findByStatus(status: PromotionStatus) {
    return this.db.promotionRun.findMany({
      where: { status },
    });
  }

  async findRecent(input: { status?: PromotionStatus; limit?: number } = {}) {
    const limit = input.limit ?? 50;

    return this.db.promotionRun.findMany({
      where: input.status ? { status: input.status } : undefined,
      orderBy: { updatedAt: 'desc' },
      take: limit,
      include: {
        pipeline: {
          include: {
            stages: {
              orderBy: { orderIndex: 'asc' },
            },
          },
        },
      },
    });
  }
}

File: kotlin/modules/db/src/main/kotlin/com/ffpromo/db/repositories/PromotionRunRepository.kt
package com.ffpromo.db.repositories

import com.ffpromo.contracts.PersistRunStateInput
import com.ffpromo.contracts.PromotionStatus
import com.ffpromo.db.DatabaseFactory
import com.ffpromo.db.IdGenerator
import com.ffpromo.db.models.PromotionRunRecord
import com.ffpromo.db.tables.Pipelines
import com.ffpromo.db.tables.PromotionRuns
import org.jetbrains.exposed.sql.ResultRow
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.insert
import org.jetbrains.exposed.sql.selectAll
import org.jetbrains.exposed.sql.update
import java.time.Instant

class PromotionRunRepository {
    fun create(input: CreatePromotionRunInput): PromotionRunRecord = DatabaseFactory.withTransaction {
        val pipeline = Pipelines
            .selectAll()
            .where { Pipelines.id eq input.pipelineId }
            .singleOrNull()
            ?: error("Pipeline ${input.pipelineId} not found")

        val now = Instant.now()
        val runId = IdGenerator.newId()

        PromotionRuns.insert {
            it[id] = runId
            it[pipelineId] = input.pipelineId
            it[pipelineVersion] = pipeline[Pipelines.version]
            it[flagKey] = input.flagKey
            it[status] = PromotionStatus.pending.name
            it[currentStageIndex] = 0
            it[temporalWorkflowId] = null
            it[pauseReason] = null
            it[createdAt] = now
            it[updatedAt] = now
        }

        findByIdInTx(runId) ?: error("PromotionRun $runId not found after insert")
    }

    fun updateState(input: UpdatePromotionRunStateInput): PromotionRunRecord = DatabaseFactory.withTransaction {
        val parsed = PersistRunStateInput(
            promotionRunId = input.promotionRunId,
            status = input.status,
            currentStageIndex = input.currentStageIndex,
            pauseReason = input.pauseReason,
        )

        val existing = PromotionRuns
            .selectAll()
            .where { PromotionRuns.id eq parsed.promotionRunId }
            .singleOrNull()
            ?: error("PromotionRun ${parsed.promotionRunId} not found")

        val existingStatus = PromotionStatus.valueOf(existing[PromotionRuns.status])
        val isFirstActiveTransition =
            existingStatus != PromotionStatus.active && parsed.status == PromotionStatus.active
        val temporalWorkflowId = input.temporalWorkflowId
            ?: if (isFirstActiveTransition && existing[PromotionRuns.temporalWorkflowId] == null) {
                parsed.promotionRunId
            } else {
                null
            }

        val now = Instant.now()
        PromotionRuns.update({ PromotionRuns.id eq parsed.promotionRunId }) {
            it[status] = parsed.status.name
            parsed.currentStageIndex?.let { stageIndex ->
                it[currentStageIndex] = stageIndex
            }
            parsed.pauseReason?.let { reason ->
                it[pauseReason] = reason
            }
            if (temporalWorkflowId != null) {
                it[PromotionRuns.temporalWorkflowId] = temporalWorkflowId
            }
            if (parsed.status == PromotionStatus.active && existing[PromotionRuns.startedAt] == null) {
                it[startedAt] = now
            }
            if (parsed.status == PromotionStatus.completed || parsed.status == PromotionStatus.aborted) {
                it[completedAt] = now
            }
            it[updatedAt] = now
        }

        findByIdInTx(parsed.promotionRunId) ?: error("PromotionRun ${parsed.promotionRunId} not found after update")
    }

    fun findById(id: String): PromotionRunRecord? = DatabaseFactory.withTransaction {
        findByIdInTx(id)
    }

    fun findByStatus(status: PromotionStatus): List<PromotionRunRecord> = DatabaseFactory.withTransaction {
        PromotionRuns
            .selectAll()
            .where { PromotionRuns.status eq status.name }
            .map { toRecord(it) }
    }

    private fun findByIdInTx(id: String): PromotionRunRecord? =
        PromotionRuns
            .selectAll()
            .where { PromotionRuns.id eq id }
            .singleOrNull()
            ?.let { toRecord(it) }

    private fun toRecord(row: ResultRow): PromotionRunRecord =
        PromotionRunRecord(
            id = row[PromotionRuns.id],
            pipelineId = row[PromotionRuns.pipelineId],
            pipelineVersion = row[PromotionRuns.pipelineVersion],
            flagKey = row[PromotionRuns.flagKey],
            status = PromotionStatus.valueOf(row[PromotionRuns.status]),
            currentStageIndex = row[PromotionRuns.currentStageIndex],
            temporalWorkflowId = row[PromotionRuns.temporalWorkflowId],
            pauseReason = row[PromotionRuns.pauseReason],
            startedAt = row[PromotionRuns.startedAt],
            completedAt = row[PromotionRuns.completedAt],
            createdAt = row[PromotionRuns.createdAt],
            updatedAt = row[PromotionRuns.updatedAt],
        )
}

data class CreatePromotionRunInput(
    val pipelineId: String,
    val flagKey: String,
)

data class UpdatePromotionRunStateInput(
    val promotionRunId: String,
    val status: PromotionStatus,
    val currentStageIndex: Int? = null,
    val pauseReason: String? = null,
    val temporalWorkflowId: String? = null,
)
